pkg/storage/chunk/cache/lru_cache.go

```diff
@@ -0,0 +1,249 @@
+package cache
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"sync"
+	"unsafe"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/loki/pkg/logqlmodel/stats"
+	util_log "github.com/grafana/loki/pkg/util/log"
+	lru "github.com/hashicorp/golang-lru/simplelru"
+	"github.com/oklog/ulid"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type codec string
+
+const (
+	codecHeaderSnappy             codec = "dvs" // As in "diff+varint+snappy".
+	codecHeaderSnappyWithMatchers codec = "dm"  // As in "dvs+matchers"
+)
+
+var DefaultLRUCacheConfig = LRUCacheConfig{
+	MaxSizeBytes: "250MB",
+}
+
+const maxInt = int(^uint(0) >> 1)
+
+const (
+	stringHeaderSize = 8
+	sliceHeaderSize  = 16
+)
+
+var ulidSize = uint64(len(ulid.ULID{}))
+
+type LRUCacheConfig struct {
+	MaxSizeBytes string `yaml:"max_size_bytes"`
+
+	Enabled bool `yaml:"enabled"`
+}
+
+// RegisterFlagsWithPrefix adds the flags required to config this to the given FlagSet
+func (cfg *LRUCacheConfig) RegisterFlagsWithPrefix(prefix, description string, f *flag.FlagSet) {
+	f.StringVar(&cfg.MaxSizeBytes, prefix+"fifocache.max-size-bytes", "500MB", description+"Maximum memory size of the cache in bytes. A unit suffix (KB, MB, GB) may be applied.")
+}
+
+func (cfg *LRUCacheConfig) Validate() error {
+	_, err := parsebytes(cfg.MaxSizeBytes)
+	return err
+}
+
+type LRUCache struct {
+	cacheType stats.CacheType
+
+	done chan struct{}
+
+	mtx sync.Mutex
+
+	logger           log.Logger
+	lru              *lru.LRU
+	maxSizeBytes     uint64
+	maxItemSizeBytes uint64
+
+	evicted     *prometheus.CounterVec
+	requests    *prometheus.CounterVec
+	hits        *prometheus.CounterVec
+	totalMisses prometheus.Counter
+	added       *prometheus.CounterVec
+	current     *prometheus.GaugeVec
+	bytesInUse  prometheus.Gauge
+	overflow    *prometheus.CounterVec
+}
+
+func NewLRUCache(name string, cfg LRUCacheConfig, reg prometheus.Registerer, logger log.Logger, cacheType stats.CacheType) (*LRUCache, error) {
+	util_log.WarnExperimentalUse(fmt.Sprintf("In-memory (LRU) cache - %s", name), logger)
+
+	maxSizeBytes, _ := parsebytes(cfg.MaxSizeBytes)
+
+	c := &LRUCache{
+		cacheType: cacheType,
+
+		maxSizeBytes: maxSizeBytes,
+		logger:       logger,
+
+		done: make(chan struct{}),
+	}
+
+	c.totalMisses = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Namespace:   "querier",
+		Subsystem:   "cache",
+		Name:        "misses_total",
```

Comment:
Please check the metrics instantiation for consistent usage of Namespace, Subsystem, Name field usages.

Reply:
Agreed. I checked after your suggestion and I made it more consistent with the existing FIFO cache, but FYI: it isn't super consistent to other caches (like redis/memcached).

I also removed all references to "index" and "querier" from this implementation since this LRU cache can be used by any Loki subsystem, not necessarily the index.